Dependency injection container

aespasa opened the pull request:
close #10

aespasa committed:
Work in progress

aespasa committed:
Implementing router v1.0

aespasa committed:
Implementing router v1.1

aespasa committed:
Implementing router

Now accept requirements as args (e.g: /users/{id:\d+}).

Request and Response are send automatically if the controller (callable) use them.

Configuration file is now a php one instead of a json.

Rewrite tests (not finished yet).

urlGenerator is not implemented yet.

aespasa committed:
Implement the router and routes.

Implement UrlGeneratorInterface.

Implement all tests in RouterTest.php, it must be run after all modifications.

The index file is the entry point. It generates the router then loads the routes inside it. Next, the router->call() is called with the Request object. If a route matches, its controller is called and it returns a Response. After all, the response is sent from the index.

When loading a route, we can pass parameters and expected regex, e.g: '/users/{id:\d+}' will look only for digits as the id parameter. If the parameter is incorrect, it returns a RouteNotFoundException.

aespasa committed:
Refactor to implement I(nterface) S(egregation) P(rinciple)

aespasa committed:
Working on dependency injection container

aespasa committed:
Implements dependency injection container and tests

We can retrieve our services automatically with the $container->get() method. If the service needs dependencies, they are also autowired.

When requesting a service from the container, it can be the same instance for all call, or a new one by specifying the $container->getDefinition(MyService::class)->setShared(false) before asking for it

Create Kernel class used in the web application entrypoint (public/index.php)

aespasa committed:
Merge branch 'main' into dependency-injection-container

aespasa committed:
Fix tests

TBoileau commented on src/Kernel.php:
```diff
@@ -0,0 +1,69 @@
+        private ContainerInterface $container,
+        private array $config = [],
+    ) {
+        $this->config['routes'] = include_once '../config/routes.php';
```
On peut le mettre dans init

TBoileau reviewed (changes requested):
Supprimer le dossier Container

aespasa committed:
Add OptionResolver

Refactor some code and replace deprecated classes

Create a ConfigFactory but I would like to replace it by the OptionResolver if it's possible.

aespasa committed:
Add DIC, update the kernel, route and router.

The order of the routes definitions is no longer important. The response is not sent by the router anymore, it's now the kernel that loads and call the router, get and call a callable provided by the router, then send a response.

Autowiring was added to the controllers methods through the kernel->run() method, using the DIC.

Refactor tests.
